typescript/web/src/components/app-lifecycle-manager/welcome-modal/welcome-modal.tsx

```diff
@@ -1,185 +1,357 @@
-
-const GithubIcon = chakra(RiGithubFill);
+  checkServiceWorkerReady,
+  messageNoWindow,
+} from "../../../utils/check-service-worker";
 
-export const WelcomeModal = ({
+const tutorialDatasetFirstImageUrl =
+  "/local/datasets/tutorial-dataset/images/2bbbf664-5810-4760-a10f-841de2f35510";
+
+type WelcomeModalParam =
+  | undefined // Default: Stays close if user already onboarded, else open asap
+  | "open" // Force it to open asap
+  | "closed"; // Force it to be closed and never open
+
+export const getDatasetsQuery = gql`
+  query getDatasets {
+    datasets {
+      id
+      name
+    }
+  }
+`;
+
+export const createDemoDatasetQuery = gql`
+  mutation createDemoDataset {
+    createDemoDataset {
+      id
+      name
+      images(first: 1) {
+        id
+        url
+      }
+      imagesAggregates {
+        totalCount
+      }
+      labelsAggregates {
+        totalCount
+      }
+      labelClassesAggregates {
+        totalCount
+      }
+    }
+  }
+`;
+
+const performWelcomeWorkflow = async ({
+  router,
   isServiceWorkerActive,
+  setIsServiceWorkerActive,
+  client,
+  setParamModalWelcome,
+  setIsLoadingWorkerAndDemo,
+  setBrowserError,
 }: {
+  router: NextRouter;
   isServiceWorkerActive: boolean;
+  setIsServiceWorkerActive: (state: boolean) => void;
+  setParamModalWelcome: (
+    state: WelcomeModalParam,
+    updateType: "replaceIn"
+  ) => void;
+  setBrowserError: (state: Error) => void;
+  setIsLoadingWorkerAndDemo: (state: boolean) => void;
+  client: ApolloClient<{}>;
 }) => {
+  try {
+    if (isServiceWorkerActive) {
+      return;
+    }
+
+    setParamModalWelcome(undefined, "replaceIn");
+
+    setIsLoadingWorkerAndDemo(true);
+
+    try {
+      await checkServiceWorkerReady();
+    } catch (e) {
+      if (e.message === messageNoWindow) {
+        // In Next JS SSR, the window is not available.
+        return;
+      }
+      if (process.env.STORYBOOK) {
+        // In Storybook, the Service Worker is not available.
+        return;
+      }
+      throw e;
+    }
+    setIsServiceWorkerActive(true);
+
+    const { data: getDatasetsResult } = await client.query({
+      query: getDatasetsQuery,
+    });
+
+    const demoDataset =
+      getDatasetsResult?.datasets == null
+        ? undefined
+        : getDatasetsResult?.datasets.filter(
+            (dataset: DatasetType) => dataset.name === "Demo dataset"
+          )?.[0] ?? undefined;
+
+    if (!demoDataset) {
+      // eslint-disable-next-line @typescript-eslint/no-unused-vars
+      const { data: createDemoDatasetResult, errors: createDemoDatasetErrors } =
+        await client.mutate({
+          mutation: createDemoDatasetQuery,
+          refetchQueries: [
+            "getDatasets",
+            "getDatasetById",
+            "getAllImagesOfADataset",
+            "getDataset",
+            "countLabelsOfDataset",
+            "getLabelClassesOfDataset",
+            "image",
+            "getImageLabels",
+          ],
```

Comment:
This looks error-prone. I am afraid we will forget to add a query here one day, or that we will rename one without seeing that the name was important.

Can't we perform a cache update rather than this?

Reply:
We discussed this with @frinyvonnick , but this is way less error prone (and 10x shorter) than a cache update (sadly). Cache updates are huge and as coupled to other pages as this.

Reply:
We can discuss during DSM